examples/ota-requestor-app/esp32/main/BDXDownloader.cpp

```diff
@@ -0,0 +1,155 @@
+#include <messaging/Flags.h>
+#include <protocols/bdx/BdxTransferSession.h>
+
+using namespace chip::bdx;
+
+void BdxDownloader::SetInitialExchange(chip::Messaging::ExchangeContext * ec)
+{
+    mExchangeCtx = ec;
+}
+
+void BdxDownloader::SetCallbacks(BdxDownloaderCallbacks callbacks)
+{
+    mOnBlockReceivedCallback    = callbacks.onBlockReceived;
+    mOnTransferCompleteCallback = callbacks.onTransferComplete;
+    mOnTransferFailedCallback   = callbacks.onTransferFailed;
+}
+
+void BdxDownloader::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (event.EventType != TransferSession::OutputEventType::kNone)
+    {
+        ChipLogDetail(BDX, "OutputEvent type: %s", event.ToString(event.EventType));
+    }
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kNone:
+        if (mIsTransferComplete)
+        {
+            if (mOnTransferCompleteCallback != nullptr && mOnTransferCompleteCallback->mCall != nullptr)
+            {
+                mOnTransferCompleteCallback->mCall(mOnTransferCompleteCallback->mContext);
+            }
+            else
+            {
+                ChipLogError(BDX, "onTransferComplete Callback not set");
+            }
+            mTransfer.Reset();
+            mIsTransferComplete = false;
+        }
+        break;
+    case TransferSession::OutputEventType::kMsgToSend: {
+        chip::Messaging::SendFlags sendFlags;
+        VerifyOrReturn(mExchangeCtx != nullptr, ChipLogError(BDX, "mExchangeContext is null, cannot proceed"));
+        if (event.msgTypeData.MessageType == static_cast<uint8_t>(MessageType::ReceiveInit))
+        {
+            sendFlags.Set(chip::Messaging::SendMessageFlags::kFromInitiator);
+        }
+        if (event.msgTypeData.MessageType != static_cast<uint8_t>(MessageType::BlockAckEOF))
+        {
+            sendFlags.Set(chip::Messaging::SendMessageFlags::kExpectResponse);
+        }
+        err = mExchangeCtx->SendMessage(event.msgTypeData.ProtocolId, event.msgTypeData.MessageType, std::move(event.MsgData),
+                                        sendFlags);
+        VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(BDX, "SendMessage failed: %" CHIP_ERROR_FORMAT, err.Format()));
+        break;
+    }
+    case TransferSession::OutputEventType::kAcceptReceived:
+        VerifyOrReturn(CHIP_NO_ERROR == mTransfer.PrepareBlockQuery(), ChipLogError(BDX, "PrepareBlockQuery failed"));
+        break;
+    case TransferSession::OutputEventType::kBlockReceived: {
+        ChipLogDetail(BDX, "Got block length %zu", event.blockdata.Length);
+
+        // TODO: while convenient, we should not do a synchronous block write in our example application - this is bad practice
+        if (mOnBlockReceivedCallback != nullptr && mOnBlockReceivedCallback->mCall != nullptr)
+        {
+            mOnBlockReceivedCallback->mCall(mOnBlockReceivedCallback->mContext, event.blockdata);
```

Comment:
Is this a synchronous call? If not, we should call back into `BdxDownloader` once it's done. Something like `OnBlockProcessed()` which then contains the rest of the logic in this function.